File: src/api/java/org/models/organization/relation/manager/HasManager.java
package org.models.organization.relation.manager;

import java.util.Set;

import org.models.organization.entity.Agent;
import org.models.organization.entity.Attribute;
import org.models.organization.id.UniqueId;
import org.models.organization.relation.Has;

/**
 * The {@linkplain HasManager} interface defines the APIs for managing {@linkplain Has}.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface HasManager {
	/**
	 * Creates a has relation between an {@linkplain Agent} and an {@linkplain Attribute} in this {@linkplain HasManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 * @param value
	 *            the <code>double</code> value.
	 */
	void addHas(UniqueId<Agent> agentId, UniqueId<Attribute> attributeId, double value);

	/**
	 * Returns a set of {@linkplain Attribute}s that is had by an {@linkplain Agent} in this {@linkplain HasManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 * @return a set of {@linkplain Attribute}s.
	 */
	Set<Attribute> getHas(UniqueId<Agent> id);

	/**
	 * Returns a set of {@linkplain Agent}s that has an {@linkplain Attribute} in this {@linkplain HasManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @return a set of {@linkplain Agent}s.
	 */
	Set<Agent> getHadBy(UniqueId<Attribute> id);

	/**
	 * Returns the <code>double</code> value of the has relation between an {@linkplain Agent} and an {@linkplain Attribute} in this {@linkplain HasManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 * @return the <code>double</code> value if the has relation exists, <code>null</code> otherwise.
	 */
	Double getHasValue(UniqueId<Agent> agentId, UniqueId<Attribute> attributeId);

	/**
	 * Sets a new <code>double</code> value for the needs relation between an {@linkplain Agent} and an {@linkplain Attribute} in this {@linkplain HasManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 * @param value
	 *            the new <code>double</code> value.
	 */
	void setHasValue(UniqueId<Agent> agentId, UniqueId<Attribute> attributeId, double value);

	/**
	 * Removes a has relation between an {@linkplain Agent} and an {@linkplain Attribute} from this {@linkplain HasManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 */
	void removeHas(UniqueId<Agent> agentId, UniqueId<Attribute> attributeId);

	/**
	 * Removes all has relations from this {@linkplain HasManager}.
	 */
	void removeAllHas();
}

File: chazm-api/src/main/java/runtimemodels/chazm/api/relation/manager/HasManager.kt
package runtimemodels.chazm.api.relation.manager

import runtimemodels.chazm.api.entity.Agent
import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.id.AgentId
import runtimemodels.chazm.api.id.AttributeId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.api.relation.Has

/**
 * The [HasManager] interface defines the APIs for managing [Has].
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface HasManager {
    /**
     * Creates a has relation between an [Agent] and an [Attribute] in this [HasManager].
     *
     * @param agentId     the [UniqueId] that represents the [Agent].
     * @param attributeId the [UniqueId] that represents the [Attribute].
     * @param value       the `double` value.
     */
    fun addHas(agentId: AgentId, attributeId: AttributeId, value: Double)

    /**
     * Returns a set of [Attribute]s that is had by an [Agent] in this [HasManager].
     *
     * @param id the [UniqueId] that represents the [Attribute].
     * @return a set of [Attribute]s.
     */
    fun getHas(id: AgentId): Set<Attribute>

    /**
     * Returns a set of [Agent]s that has an [Attribute] in this [HasManager].
     *
     * @param id the [UniqueId] that represents the [Agent].
     * @return a set of [Agent]s.
     */
    fun getHadBy(id: AttributeId): Set<Agent>

    /**
     * Returns the `double` value of the has relation between an [Agent] and an [Attribute] in this [HasManager].
     *
     * @param agentId     the [UniqueId] that represents the [Agent].
     * @param attributeId the [UniqueId] that represents the [Attribute].
     * @return the `double` value if the has relation exists, `null` otherwise.
     */
    fun getHasValue(agentId: AgentId, attributeId: AttributeId): Double?

    /**
     * Sets a new `double` value for the needs relation between an [Agent] and an [Attribute] in this [HasManager].
     *
     * @param agentId     the [UniqueId] that represents the [Agent].
     * @param attributeId the [UniqueId] that represents the [Attribute].
     * @param value       the new `double` value.
     */
    fun setHasValue(agentId: AgentId, attributeId: AttributeId, value: Double)

    /**
     * Removes a has relation between an [Agent] and an [Attribute] from this [HasManager].
     *
     * @param agentId     the [UniqueId] that represents the [Agent].
     * @param attributeId the [UniqueId] that represents the [Attribute].
     */
    fun removeHas(agentId: AgentId, attributeId: AttributeId)

    /**
     * Removes all has relations from this [HasManager].
     */
    fun removeAllHas()
}
